src/common/static_ptr.h

```diff
@@ -33,74 +33,91 @@ namespace _mem {
 enum class op {
   copy, move, destroy, size
 };
-template<typename T>
+// B is the base pointer type of the class that first created the
+// function/datum. It's only used to control copy/move support and
+// noexcept annotation. Since the ward in static_ptr_base only allows
+// moves/copies from more permissive to less, we should never end up
+// with a situation where static_ptr thinks we should be able to copy
+// or move something that can't unless people abuse
+// reinterpret_pointer_cast. But it's a reinterpreting cast. They know
+// what they're getting into.
+template<typename B, typename T>
 static std::size_t op_fun(op oper, void* p1, void* p2)
-{
+  noexcept((!std::is_copy_constructible_v<B> ||
+	    std::is_nothrow_copy_constructible_v<B>) &&
+	   (!std::is_move_constructible_v<B> ||
+	    std::is_nothrow_move_constructible_v<B>)) {
   auto me = static_cast<T*>(p1);
 
-  switch (oper) {
-  case op::copy:
-    // One conspicuous downside is that immovable/uncopyable functions
-    // kill compilation right here, even if nobody ever calls the move
-    // or copy methods. Working around this is a pain, since we'd need
-    // four operator functions and a top-level class to
-    // provide/withhold copy/move operations as appropriate.
-    new (p2) T(*me);
-    break;
-
-  case op::move:
-    new (p2) T(std::move(*me));
-    break;
-
-  case op::destroy:
+  if constexpr (std::is_copy_constructible_v<B>) {
+    if (oper == op::copy) {
+      new (p2) T(*me);
+      return 0;
+    }
+  }
+
+  if constexpr (std::is_move_constructible_v<B>) {
+    if (oper == op::move) {
+      new (p2) T(std::move(*me));
```

Comment:
variable "me" has been moved, but since it was constructed with placement-new, I'd think the old object would need to be released. However, I'm not sure it's going to be in a valid state-- perhaps what's really needed here is std::swap(). I guess the other answer is "it's the caller's pointer, they handle it".

Reply:
I was making the assumption that anything std::moved from is (generally) required to be in an undefined but valid state.

THOUGH I think the right thing to do in this case is to have the move constructors in the wrapping static_ptr empty out the calling object completely rather than destructing it here.

Reply:
Yeah, let's see what that looks like. Probably better than action-at-a-distance.